ARROW-13033: [C++] Kernel to localize naive timestamps to a timezone (preserving clock-time)

rok opened the pull request:
This is to resolve [ARROW-13033](https://issues.apache.org/jira/browse/ARROW-13033).

jorisvandenbossche reviewed (commented):
Given all the discussion on the JIRA issue, can you a bit more clearly state what exactly you implemented and what part it covers? 
For example, the current kernel returns int64 and not timestamp with timezone?

It also seems that the implementation is converting system time (UTC epoch) to local (naive) time, and not the other way around as the docstring says.

(also, given all the confusion also on the JIRA, we should probably try to come up with a more explicit/descriptive name ...)

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -177,5 +177,17 @@ TEST(ScalarTemporalTest, TestZonedTemporalComponentExtraction) {
+  for (auto u : internal::AllTimeUnits()) {
+    auto unit_naive = timestamp(u);
+    auto unit_utc = timestamp(u, timezone);
+    CheckScalarUnary("localize", unit_naive, times_naive, unit_utc, times_utc);
```
Added a couple.

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -906,48 +906,51 @@ Temporal component extraction
++--------------------+------------+-------------------+---------------+--------+----------------------------------------------+
+| subsecond          | Unary      | Temporal          | Double        |        |                                              |
++--------------------+------------+-------------------+---------------+--------+----------------------------------------------+
+| localize           | Unary      | Temporal          | Temporal      | \(4)   | :struct:`TemporalLocalizationOptions`        |
```
Done.

rok commented:
Thanks for the review @jorisvandenbossche :).
I've gone through your feedback and done some of the boilerplate except for the correct output timestamp type. I'll try to do it over the weekend.

> (also, given all the confusion also on the JIRA, we should probably try to come up with a more explicit/descriptive name ...)

I think there is no consensus on this yet. `tz_convert`, `tz_localize` were mentioned. I'd add `tz_normalize`. Any other?

jorisvandenbossche commented:
@rok can you explain the rationale of adding a ``source_timezone`` and ``destination_timezone`` to the API? What's the reason for doing it like this, instead of having a single target timezone as argument?

rok commented:
> @rok can you explain the rationale of adding a `source_timezone` and `destination_timezone` to the API? What's the reason for doing it like this, instead of having a single target timezone as argument?

In case we start with "naive" a timestamp we don't know the source and target timezone so both need to be specified. Or am I missing something here?

jorisvandenbossche commented:
What's the difference between both timezones? In my mind this is the same time zone? If you want to convert your localized timezone-aware timestamp to another timezone after localizing, you can use another kernel to do that? (a kernel we don't have yet, but one we should also add, although this is a trivial, metadata-only change)

rok commented:
> If you want to convert your localized timezone-aware timestamp to another timezone after localizing, you can use another kernel to do that? (a kernel we don't have yet, but one we should also add, although this is a trivial, metadata-only change)

I was thinking only of the metadata change not "materialization" of a target timezone. Having another kernel to only change metadata seems redundant.

jorisvandenbossche commented:
> Having another kernel to only change metadata seems redundant.

I don't think that's redundant, as a user will regularly want to change the timezone of a timestamp column that is already localized (already has a timezone).

I am of course coming from the point of view of pandas, where those two operations are defined as two distinct methods: tz_localize to convert from "timestamp without timezone" to "timestamp with timezone" (i.e. from naive clock time to tz-aware time) and tz_convert to convert between "timestamp with timezone" with different timezones (a metadata-only change). 
From a user point of view (and also implementation-wise), those two are distinct concepts, IMO, and so I would keep them as separate kernels. (also eg R's lubridate has distinct `force_tz` and `with_tz`, respectively)

jorisvandenbossche commented:
(if we keep both source and destination timezone in the "localize" kernel, I think at least the destination timezone, if not specified, should default to be the same as the source timezone, and not "UTC", so that a user doesn't have to specify the timezone they want twice, like `localize(timestamp, source_timezone="Europe/Brussels", destination_timezone="Europe/Brussels")` to end up with a `timestamp(unit, tz="Europe/Brussels)` type column)

jorisvandenbossche commented:
I opened https://issues.apache.org/jira/browse/ARROW-13247 for the discussion whether we want a separate "change timezone" kernel

rok commented:
> > Having another kernel to only change metadata seems redundant.
> 
> I don't think that's redundant, as a user will regularly want to change the timezone of a timestamp column that is already localized (already has a timezone).
> 
> I am of course coming from the point of view of pandas, where those two operations are defined as two distinct methods: tz_localize to convert from "timestamp without timezone" to "timestamp with timezone" (i.e. from naive clock time to tz-aware time) and tz_convert to convert between "timestamp with timezone" with different timezones (a metadata-only change).
> From a user point of view (and also implementation-wise), those two are distinct concepts, IMO, and so I would keep them as separate kernels. (also eg R's lubridate has distinct `force_tz` and `with_tz`, respectively)

I suppose these are two different operations indeed. I've changed to make this only "localization" now.

I like `force_tz` as it indicates the operation is a bit "dangerous". Candidates are then:
* force_tz
* tz_localize
* tz_normalize

adamhooper commented:
Since the Arrow spec is going to refer to "LocalDateTime", "Instant" and "ZonedDateTime", how about reusing terms from the spec? Something along the lines of `local_datetime_to_zoned_datetime()`?

rok commented:
> Since the Arrow spec is going to refer to "LocalDateTime", "Instant" and "ZonedDateTime", how about reusing terms from the spec? Something along the lines of `local_datetime_to_zoned_datetime()`?

It's nice that it's idiomatic if a bit verbose :)

Let's add it to the list. If we don't get to consensus here we can always call a vote on the ML.
* force_tz
* tz_localize
* tz_normalize
* local_datetime_to_zoned_datetime

westonpace commented:
I like all except `tz_normalize`.  "normalize" just has too many meanings for me already.

jorisvandenbossche commented:
@rok I was checking out this PR, and in the process wrote python bindings for the option class to be able to test it in Python, so thought to directly push that here as well. (and at the same time also fixed (I think) the initialization of the timezone, which was causing the C++ tests to fail here as well)

And from testing, you will still need to handle the errors raised by `date.h`:

```
In [21]: arr = pa.array([pd.Timestamp("2021-03-28 02:30:00")], type=pa.timestamp("ns"))

In [22]: pc.localize(arr, options=pc.TemporalLocalizationOptions("Europe/Brussels"))
terminate called after throwing an instance of 'arrow_vendored::date::nonexistent_local_time'
  what():  2021-03-28 02:30:00.000000000 is in a gap between
2021-03-28 02:00:00 CET and
2021-03-28 03:00:00 CEST which are both equivalent to
2021-03-28 01:00:00 UTC
Aborted (core dumped)

```

rok commented:
> @rok I was checking out this PR, and in the process wrote python bindings for the option class to be able to test it in Python, so thought to directly push that here as well. (and at the same time also fixed (I think) the initialization of the timezone, which was causing the C++ tests to fail here as well)

Thanks @jorisvandenbossche! The new option initialization was indeed problematic.

rok commented:
@jorisvandenbossche I've added nonexistent and ambiguous time handling. I'll try to template calls instead of having a single general one.

What I'm not sure about now is if try/catch is the way to go here and how to return `NaN` when the timestamp is an `int64`.

Another thing: `infer` requires understanding if the array is monotonous or not (if I understand correctly) and I'd push it out of scope of this PR for now.

jorisvandenbossche commented:
> What I'm not sure about now is if try/catch is the way to go here

I think the try/catch is fine, it's the way to convert an exception (which we don't use ourselves as return values) into a Status. 

> and how to return NaN when the timestamp is an int64.

I think it should rather be a null instead of NaN (I suppose this is for the IGNORE case?), and nulls can be encoded in int64. 

I also think that the output type of the kernel should be timestamp _with_ timezone.

> Another thing: `infer` requires understanding if the array is monotonous or not (if I understand correctly) and I'd push it out of scope of this PR for now.

Yes, that's certainly something to be left/discussed for later, as it makes the kernel no longer a simple scalar kernel (where we can operate on each scalar independently element-wise), since the next scalar needs to have knowledge about the previous ones.

rok commented:
> > and how to return NaN when the timestamp is an int64.
> 
> I think it should rather be a null instead of NaN (I suppose this is for the IGNORE case?), and nulls can be encoded in int64.

Yes I was thinking about the IGNORE case. This should now return nulls but zeros are returned. Do we need to switch away from `ScalarUnaryNotNullStateful` for this?

> I also think that the output type of the kernel should be timestamp _with_ timezone.

Done. I was not sure how to get the timezone into the out type. I've made a runtime resolver now.

I'll move this out of draft to invite more discussion :).

rok commented:
[CI fails with](https://github.com/apache/arrow/pull/10610/checks?check_run_id=3045835197#step:11:175):
```
C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: ../windows/arrow-4.0.1.9000/lib/i386/libarrow.a(tz.cpp.obj):(.text+0x725): undefined reference to `_imp__CoTaskMemFree@4'
C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: ../windows/arrow-4.0.1.9000/lib/i386/libarrow.a(tz.cpp.obj):(.text.unlikely+0x3e): undefined reference to `_imp__CoTaskMemFree@4'
```

It seems to be due to `tz` object being problematic on windows (where currently tzdb is not available).
It appears that [ARROW-13033](https://issues.apache.org/jira/browse/ARROW-13033) and [ARROW-12980](https://issues.apache.org/jira/browse/ARROW-12980) are now both blocked by [ARROW-13168](https://issues.apache.org/jira/browse/ARROW-13168).

rok commented:
I removed the R bindings for now and we can reintroduce it with [ARROW-13168](https://issues.apache.org/jira/browse/ARROW-13168).

rok commented:
That did not help. [This suggests](https://github.com/HowardHinnant/date/issues/272) adding `-lole32` to mingw.

rok commented on r/configure.win:
```diff
@@ -44,7 +44,7 @@ else
 fi
 OPENSSL_LIBS="-lcrypto -lcrypt32"
-MIMALLOC_LIBS="-lbcrypt -lpsapi"
+MIMALLOC_LIBS="-lbcrypt -lpsapi -lole32"
```
Should this be somewhere else?

westonpace replied:
No, I didn't see your earlier comment so I didn't understand why you were adding this.  I get it now, sorry.

westonpace replied:
Ah, sorry, thought this was on the other PR where I was questioning this.  I will let someone with more knowledge of the build process chime in on whether this is correct or not.  I thought the TZ stuff didn't support Windows anyways?

rok replied:
No worries, thanks for responding :)
The TZ functionality of the vendored library (date.h) needs to download a timezone DB on windows to work. There's [a jira to set that up](https://issues.apache.org/jira/browse/ARROW-13168) so this will eventually need to compile on windows.

jorisvandenbossche commented:
> > > and how to return NaN when the timestamp is an int64.
> > 
> > 
> > I think it should rather be a null instead of NaN (I suppose this is for the IGNORE case?), and nulls can be encoded in int64.
> 
> Yes I was thinking about the IGNORE case. This should now return nulls but zeros are returned. Do we need to switch away from `ScalarUnaryNotNullStateful` for this?

If it makes the implementation more complicated (as it would no longer simply propagate nulls), I think it is fine to also leave this option for a future enhancement

rok commented:
> > > > and how to return NaN when the timestamp is an int64.
> > > 
> > > I think it should rather be a null instead of NaN (I suppose this is for the IGNORE case?), and nulls can be encoded in int64.
> > 
> > Yes I was thinking about the IGNORE case. This should now return nulls but zeros are returned. Do we need to switch away from `ScalarUnaryNotNullStateful` for this?
> 
> If it makes the implementation more complicated (as it would no longer simply propagate nulls), I think it is fine to also leave this option for a future enhancement

I suppose it would. Shall we leave this PR as is then?

jorisvandenbossche commented:
> Shall we leave this PR as is then?

I would then remove this option (if we keep it in the PR, it should do the correct thing IMO, while not it is still returning 0 (i.e. 1970-01-01) instead I think?)

jorisvandenbossche commented:
Can you also handle the case of an unknown / wrong timezone? (I thought I already commented about that before, but don't directly see it ;)) Currently when using such a timezone, you get a segfault:

```
In [7]: pc.tz_localize(arr, timezone="Europe/Brussels", nonexistent="shift_forward").type
Out[7]: TimestampType(timestamp[ns, tz=Europe/Brussels])

In [8]: pc.tz_localize(arr, timezone="Europe/Brusselsss", nonexistent="shift_forward")
terminate called after throwing an instance of 'std::runtime_error'
  what():  Europe/Brusselsss not found in timezone database
Aborted (core dumped)
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -321,6 +341,77 @@ struct Nanosecond {
+          return get_local_time<T, Arg0>(arg, arrow_vendored::date::choose::earliest);
+        }
+        case TemporalLocalizationOptions::Ambiguous::LAST: {
+          return get_local_time<T, Arg0>(arg, arrow_vendored::date::choose::latest);
```
For `nonexistent`, we use "shift_backward" and "shift_forward" for earliest/latest, and here we use "first" and "last" for the `ambiguous` keyword. Shall we make this consistent?  (and could then also use the terminology of date.h)

rok replied:
Yeah, that makes sense. I've made names consistent with `date.h`. See the latest commit.

rok commented:
Thanks for the review @jorisvandenbossche! :)

> > Shall we leave this PR as is then?
> 
> I would then remove this option (if we keep it in the PR, it should do the correct thing IMO, while not it is still returning 0 (i.e. 1970-01-01) instead I think?)

Removed the `IGNORE` option and made [ARROW-13347](https://issues.apache.org/jira/browse/ARROW-13347) to discuss/remember.

> Can you also handle the case of an unknown / wrong timezone?

Added but not at `TemporalLocalizationOptions` initialization but in [the TzLocalize exec](https://github.com/apache/arrow/pull/10610/files#diff-53850537a2f3f7977473b30241fdd3ddb35aa97269284c06471d9b0c599e2c59R86) since state can be returned from there.

I think this is ready for another review.

rok commented:
@pitrou - I've rebased this and it's ready for review.

jorisvandenbossche commented:
We still have the naming issue (or bikeshedding) to resolve. As a reference, some names used by other systems that I am aware of:
* Pandas: [tz_localize](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.tz_localize.html)
* R's lubridate: [force_tz](https://lubridate.tidyverse.org/reference/force_tz.html)
* Java: [atZone](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html#atZone-java.time.ZoneId-)
* Java Joda: [toDateTime](https://www.joda.org/joda-time/apidocs/org/joda/time/LocalDateTime.html#toDateTime-org.joda.time.DateTimeZone-)
* Postgresql: [AT TIME ZONE](https://www.postgresql.org/docs/13/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT)

The proposals that have been floated around above:

* `tz_localize`
* `force_tz`
* `local_datetime_to_zoned_datetime`

I left out `tz_normalize` (like Weston mentioned, "normalize" already has too many meanings, it also doesn't give any indication to me whether it's adding or removing a timezone ..). 
I personally like Adam's suggestion of being more explicit (although resulting in a longer name) and re-using the terminology that was now added to the format spec to clarify the meaning. I would maybe only use "timestamp" instead of "datetime", since the actual types in arrow that this kernel accepts are called that way. 
With that, I would do a proposal of calling it `timestamp_local_to_zoned` ?

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -570,6 +586,71 @@ struct Strftime {
+    } catch (const arrow_vendored::date::nonexistent_local_time& e) {
+      switch (options.nonexistent) {
+        case TemporalLocalizationOptions::Nonexistent::NONEXISTENT_RAISE: {
+          *st = Status::Invalid("Nonexistent: ", e.what());
```
```suggestion
          *st = Status::Invalid("Nonexistent timestamp: ", e.what());
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -570,6 +586,71 @@ struct Strftime {
+    } catch (const arrow_vendored::date::ambiguous_local_time& e) {
+      switch (options.ambiguous) {
+        case TemporalLocalizationOptions::Ambiguous::AMBIGUOUS_RAISE: {
+          *st = Status::Invalid("Ambiguous: ", e.what());
```
```suggestion
          *st = Status::Invalid("Ambiguous timestamp: ", e.what());
```

pitrou commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -269,6 +270,31 @@ struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
+  enum Ambiguous { AMBIGUOUS_RAISE, AMBIGUOUS_EARLIEST, AMBIGUOUS_LATEST };
+
+  /// How to handle local times that do not exists due to DST shifts.
+  enum Nonexistent { NONEXISTENT_RAISE, NONEXISTENT_EARLIEST, NONEXISTENT_LATEST };
```
@bkietz @jorisvandenbossche  Any thoughts about naming? Should these be called `AmbiguousHandling` and `NonExistentHandling`?

pitrou commented:
As for naming the function, perhaps "localize_timestamp"? Most important is to have a good explanatory docstring.

jorisvandenbossche commented:
Although it's named like this in pandas, I am a bit hesitant to use "localize". I seem to recall that eg @bkietz interpreted that as the reverse operation (from zoned to naive/local), and in general it seemed to cause confusion in the discussion around this. 

And while typing that, I realize that it would indeed be nice to have a kernel for the reverse as well (zoned to local), so the question is how to name that if we call this "localize_timestamp" (while for "timestamp(_local)_to_zoned" there is a obvious "timestamp(_zoned)_to_local" equivalent naming option)

pitrou commented:
"localize_naive_timestamp" perhaps?

pitrou commented:
The problem is we don't have any clear terminology here. "zoned" doesn't mean anything to me.

jorisvandenbossche commented:
I was following the new terminology at 

https://github.com/apache/arrow/blob/895f85fd32782a4447db74e4413e54a4f7128809/format/Schema.fbs#L221-L247

pitrou commented:
I don't read that as new terminology, rather pointers to terminology used in some other pieces of software.
(also I find that explanation actually a bit intimidating and confusing)

jorisvandenbossche commented:
Next name suggestion: `to_timestamp_with_time_zone`. 
It doesn't use any specific terminology like "local" or "zoned", but only refers to the actual parameter of the timestamp type: whether it has the time zone set or not. 

> I don't read that as new terminology, rather pointers to terminology used in some other pieces of software.
(also I find that explanation actually a bit intimidating and confusing)

The edits in https://github.com/apache/arrow/pull/10629 where meant to _clarify_ the text after the discussion on the mailing list (and not make it even more confusing :)), so if that's not the case, I think your feedback would be welcome.

rok commented:
@jorisvandenbossche I think most of the names we're discussing are great. I would slightly prefer an 'old' name as it confuses slightly less users :).
I'm just going through the feedback now and will push a commit today.

rok commented:
@jorisvandenbossche ok, this is ready for another review.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -840,6 +967,13 @@ const FunctionDoc strftime_doc{
 
+const FunctionDoc tz_localize_doc{
+    "TzLocalize converts timestamps to UTC keeping the original timezone information",
+    ("TzLocalize converts timestamps from arbitrary timezone to UTC.\n"
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -570,6 +586,71 @@ struct Strftime {
 #endif
 
+// ----------------------------------------------------------------------
+// Convert timestamps from arbitrary timezone to UTC
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -161,6 +162,22 @@ struct TemporalComponentExtract
+    const TemporalLocalizationOptions& options = TemporalLocalizationState::Get(ctx);
+    const auto& timezone = GetInputTimezone(batch.values[0]);
+    if (!timezone.empty() && options.timezone != timezone) {
+      return Status::Invalid("Input and output timezones do not match: ", timezone,
```
Yeah, that's a good point! I made this so that input timestamp array has to be timezone-less.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -1029,5 +1055,19 @@ ARROW_EXPORT Result<Datum> Subsecond(const Datum& values, ExecContext* ctx = NUL
 ARROW_EXPORT Result<Datum> Strftime(const Datum& values, StrftimeOptions options,
                                     ExecContext* ctx = NULLPTR);
 
+/// \brief Converts timestamps from one timezone to another for each element of `values`
```
Done.

rok commented:
Name proposals so far:
* `tz_localize`
* `force_tz`
* `local_datetime_to_zoned_datetime`
* `timestamp_to_zoned`
* `timestamp_to_local`
* `timestamp_local_to_zoned`
* `timestamp_local_at_zone`
* `localize_timestamp`
* `localize_naive_timestamp`
* `to_timestamp_with_time_zone`

I kinda like '..._at_zone' format because it makes me think the timestamp will be interpreted as in a timezone rather then converted to.

pitrou commented:
I'm going to suggest `assume_timezone`.

adamhooper commented:
@rok Thanks for asking! I prefer `assume_timezone`, too.

pitrou commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -278,6 +279,31 @@ struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
+
+  /// Timezone and timezone name to convert timestamp from
+  std::string timezone;
+  const arrow_vendored::date::time_zone* tz;
```
I don't think this should be here. Is it deliberate or just a leftover?

rok replied:
Oh. Probably a bad rebase.

rok replied:
Removed.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -571,6 +592,73 @@ struct Strftime {
+        }
+      }
+    }
+    *st = Status::UnknownError("Unknown error.");
```
I don't really understand what this is for. Presumably this should be unreachable?

rok replied:
I think it was just to satisfy the compiler.

rok replied:
Removed the status but kept the `return 0` to keep he compiler happy. Please check.

rok commented:
CI fail seems unrelated.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -383,7 +384,96 @@ TEST_F(ScalarTemporalTest, DayOfWeek) {
+  std::string timezone = "Europe/Warsaw";
+  const char* times = R"(["2015-03-29 02:30:00", "2015-03-29 03:30:00"])";
+  const char* times_earliest = R"(["2015-03-29 01:00:00", "2015-03-29 01:30:00"])";
+  const char* times_latest = R"(["2015-03-29 01:00:00", "2015-03-29 01:30:00"])";
```
Am I misreading, or is `times_latest` equal to `times_earliest`?

pitrou replied:
I guess that the `date` library chooses the existing (UTC) instant that encloses the invalid instant "2015-03-29 02:30:00 Europe/Paris`. But having two option values that produce the same results doesn't seem useful at all.

rok replied:
Yeah, indeed https://howardhinnant.github.io/date/tz.html#nonexistent_local_time

rok replied:
I'll see if we can make it more like expected.

rok replied:
Now the behavior is more like shift forward and shift backward.

pitrou replied:
Uh, sorry, that's what not I meant :-) I think the behaviour was ok, it just that there shouldn't be two different option values that produce exactly the same result.

rok replied:
Well maybe we can keep the original behavior as `infer` and have the two new ones as `shift_forward` and `shift_forward`. It needs a new name for sure.
I'm borrowing from [pandas tz_localize](https://pandas.pydata.org/docs/reference/api/pandas.Series.tz_localize.html).

pitrou replied:
Well, that depends, do we want to copy Pandas here?
Look at the results in the Pandas example: "2015-03-29 01:59:59.999999999+01:00" and "2015-03-29 03:00:00+02:00" are separated by a single _nanosecond_. Is this behaviour useful? @jorisvandenbossche 

Also, note that `shift_forward` already corresponds to the original `date` library's behaviour. So we don't need an additional option values :-)

rok replied:
> Look at the results in the Pandas example: "2015-03-29 01:59:59.999999999+01:00" and "2015-03-29 03:00:00+02:00" are separated by a single _nanosecond_. Is this behaviour useful? @jorisvandenbossche

I'm sure there are is someone out there who care about nanoseconds :)).

My thought here is - the input timestamp is half past full hour locally. The local hour doesn't exist, so the half past is probably a measurement mistake and what's was measured is half past an hour later or earlier. The third option (infer) is to ignore the half past and go to the moment the DST shift occurred.

```
earliest -> 2015-03-29 02:30:00 (Warsaw) -> 2015-03-29 00:30:00 (UTC)
infer -> 2015-03-29 02:30:00 (Warsaw) -> 2015-03-29 01:00:00 (UTC)
latest -> 2015-03-29 01:30:00 (Warsaw) -> 2015-03-29 01:30:00 (UTC)
```

Visual aid for the discussion (not exactly the same zones):
![](https://files.gitter.im/HowardHinnant/date/fM7T/EuropeMoscow.jpeg)

We can also just park this into [ARROW-13347](https://issues.apache.org/jira/browse/ARROW-13347).

pitrou replied:
> My thought here is - the input timestamp is half past full hour locally. The local hour doesn't exist, so the half past is probably a measurement mistake and what was measured is half past an hour later or earlier. 

This corresponds neither to the Pandas behaviour nor to the `date` library behaviour. Is there any system that offers such an option? Saying "the half past is probably a measurement mistake and what was measured is half past an hour later or earlier" is quite a wild guess...

pitrou replied:
Bottom line, I think we can simply implement the Pandas semantics while keeping the name "earliest" and "latest".

rok replied:
> This corresponds neither to the Pandas behaviour nor to the `date` library behaviour. Is there any system that offers such an option? Saying "the half past is probably a measurement mistake and what was measured is half past an hour later or earlier" is quite a wild guess...

Well if the user wants to do it it's not guessing :)
Pandas is close with the `nonexistent=pd.Timedelta('1H')` but yeah, it is probably to niche to have here.

rok replied:
> Bottom line, I think we can simply implement the Pandas semantics while keeping the name "earliest" and "latest".

Sounds good, will do.

rok replied:
Pushed a pandas-like change. It's a bit annoying that output depends on the unit but ok ..

pitrou replied:
Thank you!

pitrou commented:
Ok, I've pushed some minor changes and will merge once CI is green. Thanks a lot for this @rok !

rok commented:
Let me rebase that.

pitrou commented:
@rok I'm on it already.

rok commented:
Thanks @jorisvandenbossche @pitrou @adamhooper @westonpace !
